### PR TITLE
Navigation: Suggest maps to download

### PR DESCRIPTION
**Pull request by androiddevkotlin**
https://github.com/osmandapp/OsmAnd/issues/9144

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/routing/RouteProvider.java [hunk 1]
```diff
@@ -76,6 +76,7 @@
 	private static final int MIN_DISTANCE_FOR_INSERTING_ROUTE_SEGMENT = 60;
 	private static final int ADDITIONAL_DISTANCE_FOR_START_POINT = 300;
 	private static final int MIN_STRAIGHT_DIST = 50000;
+	private static final long DELAY = 60000;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteProvider.java:79]: bad constant name

File: OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java [hunk 2]
```diff
@@ -254,13 +254,16 @@ void updateProgress(final RouteCalculationParams params) {
...
 					RouteCalculationProgress calculationProgress = params.calculationProgress;
+					long routeCalculationDuration = System.currentTimeMillis();
 					if (isRouteBeingCalculated()) {
 						if (lastTask != null && lastTask.params == params) {
 							progressRoute.updateProgress((int) calculationProgress.getLinearProgress());
 							if (calculationProgress.requestPrivateAccessRouting) {
 								progressRoute.requestPrivateAccessRouting();
 							}
-							if (!Algorithms.isEmpty(params.missingMaps)) {
+							if (!Algorithms.isEmpty(params.missingMaps) && routeCalculationDuration < params.startTimeRouteCalculation) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java:264]: weird combination of if conditions '<' and else '>'

File: OsmAnd/src/net/osmand/plus/routing/RouteProvider.java [hunk 3]
```diff
@@ -118,6 +119,10 @@ public RouteCalculationResult calculateRouteImpl(RouteCalculationParams params)
 					} else {
 						if (!suggestionsMapsProvider.isPointOnWater()) {
 							params.missingMaps = suggestedMapsOnStraightLine;
+						} else {
+							List<Location> onlinePoints = suggestionsMapsProvider.findOnlineRoutePoints();
+							params.missingMaps = suggestionsMapsProvider.getMissingMaps(onlinePoints);
+							params.startTimeRouteCalculation = System.currentTimeMillis() + DELAY;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteProvider.java:125]: variable name says "startTime" but the value equals now + DELAY

File: OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java [hunk 4]
```diff
@@ -190,8 +190,39 @@ public String getLocaleNameByFullName(String fullName, boolean includingParent)
...
+	public String getSuggestedDownloadMapName(String downloadName) {
+		final String lc = downloadName.toLowerCase();
+		if (downloadNamesToFullNames.containsKey(lc)) {
+			String fullName = downloadNamesToFullNames.get(lc);
+			return getSuggestedMapsName(fullName);
+		}
+		return downloadName.replace('_', ' ');
+	}
+
+	public String getSuggestedMapsName(String fullName) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java:210]: Unclear method copy / paste probably not needed

File: OsmAnd-java/src/main/java/net/osmand/util/MapUtils.java [hunk 5]
```diff
@@ -751,4 +756,17 @@ public static double getSqrtDistance(int startX, int startY, int endX, int endY)
 	public static double getSqrtDistance(float startX, float startY, float endX, float endY) {
 		return Math.sqrt((endX - startX) * (endX - startX) + (endY - startY) * (endY - startY));
 	}
+
+	//Looks like a method rhumbDestinationPoint. The result is also looks the same, for example: 19983.254 and 19983.266. Please, choose better method.
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/util/MapUtils.java:760]: probably not needed anymore

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java [hunk 6]
```diff
@@ -193,6 +196,11 @@
 	private boolean editButtonCollapsed;
 	private boolean addButtonCollapsing;
 	private boolean addButtonCollapsed;
+	private boolean isStartFinishIntermediatesMaps;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:199]: Both variables needed?

Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:199]: Probably name is bad cause it means only "disableNavigation"

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java [hunk 7]
```diff
@@ -193,6 +196,11 @@
...
+
+	private List<WorldRegion> missingMapsOnDirectLine;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:201]: Only 1 variable is needed like suggestedMissingMaps

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/cards/SuggestionsMapsDownloadWarningCard.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+				downloadThread.runReloadIndexFiles();
+			}
+		}
+
+		boolean downloadIndexes = app.getSettings().isInternetConnectionAvailable()
+				&& !downloadThread.getIndexes().isDownloadedFromInternet
+				&& !downloadThread.getIndexes().downloadFromInternetFailed;
+
+		List<WorldRegion> suggestedMaps = new ArrayList<>();
+		if (isMissingMapsStartFinishIntermediates) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/cards/SuggestionsMapsDownloadWarningCard.java:129]: why don't make as 1 method ?without ifs

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java [hunk 9]
```diff
@@ -79,6 +82,26 @@
 
 	public RouteCalculationResult(String errorMessage) {
 		this.errorMessage = errorMessage;
+		this.downloadMaps = null;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java:85]: Copy / paste

File: OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java [hunk 10]
```diff
@@ -345,6 +352,7 @@ public void run() {
 			}
 			RouteCalculationResult prev = routingHelper.getRoute();
 			synchronized (routingHelper) {
+				routingHelper.setRoute(res);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java:355]: Dangerous code - could seriously affect how application works!

File: OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java [hunk 11]
```diff
@@ -376,7 +384,9 @@ public void run() {
 					routeCalcError = app.getString(R.string.empty_route_calculated);
 					routeCalcErrorShort = app.getString(R.string.empty_route_calculated);
 				}
-				showMessage(routeCalcError);
+				if (res.getDownloadMaps().isEmpty()) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteRecalculationHelper.java:387]: Dangerous code - could seriously affect how application works!

File: OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java [hunk 12]
```diff
@@ -185,13 +185,25 @@ public String getLocaleNameByFullName(String fullName, boolean includingParent)
...
 				return rd.getLocaleName();
 			}
-			if (parentParent.getRegionId().equals(WorldRegion.RUSSIA_REGION_ID)) {
-				return parentParent.getLocaleName() + " " + rd.getLocaleName();
-			}
-			if (parentParent.getRegionId().equals(WorldRegion.JAPAN_REGION_ID)) {
-				return parentParent.getLocaleName() + " " + rd.getLocaleName();
+			if (parentFirst) {
+				if (parentParent.getRegionId().equals(WorldRegion.RUSSIA_REGION_ID)) {
+					return parentParent.getLocaleName() + " " + rd.getLocaleName();
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/map/OsmandRegions.java:190]: return String.format(format, parentParent.getLocaleName(), rd.getLocaleName());

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+	public List<Location> getLocationBasedOnDistanceInterval(List<Location> points) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0; i < points.size(); i++) {
+			int nextIndex = i + 1 < points.size() ? i + 1 : i;
+			mapsBasedOnPoints.add(0, points.get(i));
+			mapsBasedOnPoints.add(mapsBasedOnPoints.size(), points.get(nextIndex));
+			while (mapsBasedOnPoints.get(0).distanceTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1)) > DISTANCE) {
+				float bearing = mapsBasedOnPoints.get(0).bearingTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1));
+				Location location = movePointToDistance(mapsBasedOnPoints.get(0).getLatitude(), mapsBasedOnPoints.get(0).getLongitude(), DISTANCE, bearing);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:56]: still this method is not deleted ? we discussed it's not correct

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	public List<Location> getLocationBasedOnDistanceInterval(List<Location> points) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0; i < points.size(); i++) {
+			int nextIndex = i + 1 < points.size() ? i + 1 : i;
+			mapsBasedOnPoints.add(0, points.get(i));
+			mapsBasedOnPoints.add(mapsBasedOnPoints.size(), points.get(nextIndex));
+			while (mapsBasedOnPoints.get(0).distanceTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1)) > DISTANCE) {
+				float bearing = mapsBasedOnPoints.get(0).bearingTo(mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1));
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:55]: too many repetitions mapsBasedOnPoints.get(0) and mapsBasedOnPoints.get(mapsBasedOnPoints.size() - 1)

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+		}
+		OnlineRoutingEngine.OnlineRoutingResponse response =
+				helper.calculateRouteOnline(createInitStateEngine(), points, params.leftSide);
+		if (response != null) {
+			route = response.getRoute();
+			routeLocation.addAll(route);
+		}
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0; i < routeLocation.size(); i++) {
+			for (int j = i + 1; j < routeLocation.size(); j++) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:93]: very confusing code

File: OsmAnd/src/net/osmand/plus/download/CustomIndexItem.java [hunk 16]
```diff
@@ -66,14 +66,8 @@ public File getTargetFile(@NonNull OsmandApplication ctx) {
...
 	@Override
-	public String getVisibleName(Context ctx, OsmandRegions osmandRegions) {
-		return getVisibleName(ctx, osmandRegions, true);
-	}
-
-	@Override
-	public String getVisibleName(Context ctx, OsmandRegions osmandRegions, boolean includingParent) {
-		String name = super.getVisibleName(ctx, osmandRegions, includingParent);
-		return JsonUtils.getLocalizedResFromMap(ctx, names, name);
+	public String getVisibleName(Context ctx, OsmandRegions osmandRegions, String format) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/CustomIndexItem.java:69]: Зачем менять? нужно оставить старые методы для совместимости

File: OsmAnd/src/net/osmand/plus/download/DownloadActivityType.java [hunk 17]
```diff
@@ -347,7 +345,7 @@ public String getVisibleDescription(DownloadItem downloadItem, Context ctx) {
 		return "";
 	}
 
-	public String getVisibleName(DownloadItem downloadItem, Context ctx, OsmandRegions osmandRegions, boolean includingParent) {
+	public String getVisibleName(DownloadItem downloadItem, Context ctx, OsmandRegions osmandRegions, String format) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/DownloadActivityType.java:348]: Зачем менять? нужно оставить старые методы для совместимости

File: OsmAnd/src/net/osmand/plus/download/SelectIndexesHelper.java [hunk 18]
```diff
@@ -249,7 +249,7 @@ private void setUseMetersForAllItems(boolean useMeters) {
 
 	private void updateSelectableItem(SelectableItem selectableItem,
 	                                  DownloadItem downloadItem) {
-		selectableItem.setTitle(downloadItem.getVisibleName(app, app.getRegions(), false));
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/SelectIndexesHelper.java:252]: Зачем менять? нужно оставить старые методы для совместимости

File: OsmAnd/src/net/osmand/plus/download/ui/ItemViewHolder.java [hunk 19]
```diff
@@ -163,7 +163,7 @@ public void bindDownloadItem(final DownloadItem downloadItem, final String cityN
 		if (showTypeInName) {
 			name = downloadItem.getType().getString(context);
 		} else {
-			name = downloadItem.getVisibleName(context, context.getMyApplication().getRegions(), showParentRegionName);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/ui/ItemViewHolder.java:166]: Зачем менять? нужно оставить старые методы для совместимости

File: OsmAnd/src/net/osmand/plus/download/ui/ItemViewHolder.java [hunk 20]
```diff
@@ -513,7 +513,7 @@ private void confirmRemove(@NonNull final DownloadItem downloadItem,
 			message = context.getString(R.string.delete_number_files_question, downloadedFiles.size());
 		} else {
 			OsmandRegions regions = app.getRegions();
-			String visibleName = downloadItem.getVisibleName(context, regions);
+			String visibleName = downloadItem.getVisibleName(context, regions, "%1$s %2$s");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/ui/ItemViewHolder.java:516]: Зачем менять? нужно оставить старые методы для совместимости

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+	public SuggestionsMapsProvider(RouteCalculationParams params) {
+		this.params = params;
+	}
+
+	public boolean isPointOnWater() {
+		return pointOnWater;
+	}
+
+	public boolean checkIfObjectDownloaded(String downloadName) {
+		final String regionName = Algorithms.capitalizeFirstLetterAndLowercase(downloadName)
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:43]: Check is not correct, cause it doesn't check actual data if user has other maps for this bbox with different name

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+				+ IndexConstants.BINARY_MAP_INDEX_EXT;
+		final String roadsRegionName = Algorithms.capitalizeFirstLetterAndLowercase(downloadName) + ".road"
+				+ IndexConstants.BINARY_MAP_INDEX_EXT;
+		return params.ctx.getResourceManager().getIndexFileNames().containsKey(regionName) || params.ctx.getResourceManager().getIndexFileNames().containsKey(roadsRegionName);
+	}
+
+	public List<Location> getLocationBasedOnDistanceInterval(List<Location> points) {
+		for (int i = 0; i < points.size(); i++) {
+			int nextIndex = i + 1 < points.size() ? i + 1 : i;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:52]: weird construction with nextindex and i + 1 == points.size

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+			while (points.get(i).distanceTo(points.get(nextIndex)) > DISTANCE) {
+				Location location = calculateMidPoint(points.get(i), points.get(nextIndex));
+				points.add(nextIndex, location);
+			}
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:56]: 1. Original Points are always in place ? It's better to have separate List to return and not to modify original list! 
Especially with method called 'get' (it should not modify anything).
2. if points are distributed like every 50 meters, the method will keep all the points, which is probably too much what's needed

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		return points;
+	}
+
+	public LinkedList<Location> getStartFinishIntermediatesPoints(String locationProvider) {
+		LinkedList<Location> points = new LinkedList<>();
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:62]: LinkedList - is not clear why needed (List?) - ArrayList might be good as well

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+				if (routeLocation.get(i).distanceTo(routeLocation.get(j)) >= DISTANCE) {
+					mapsBasedOnPoints.add(routeLocation.get(j));
+					break;
+				}
+			}
+		}
+		return mapsBasedOnPoints;
+	}
+
+	private OnlineRoutingEngine createInitStateEngine() {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:140]: this probably should be part of OnlineRoutingHelper and it should be allowed to return null. This method actually should work only for CAR / BICYCLE/ FOOT  profile and it should it ignore all others

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+				suggestedMaps.addAll(maps);
+			}
+		}
+		return new ArrayList<>(suggestedMaps);
+	}
+
+	@NonNull
+	private List<Location> getOnlineLocationsBasedOnDistance(List<Location> routeLocation) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0, j; i < routeLocation.size(); i = j) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:129]: this an impossible loop with all variables messed up, it is not clear what's going on here

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java [hunk 27]
```diff
@@ -95,6 +98,25 @@ public RouteCalculationResult(String errorMessage) {
 		this.routeVisibleAngle = 0;
 	}
 
+	public RouteCalculationResult(List<WorldRegion> suggestedOfflineMaps) {
+		this.downloadMaps = suggestedOfflineMaps;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java:102]: Copy paste!

File: OsmAnd/src/net/osmand/plus/routing/RouteProvider.java [hunk 28]
```diff
@@ -130,7 +147,7 @@ public RouteCalculationResult calculateRouteImpl(RouteCalculationParams params)
 				log.error("Failed to find route ", e); //$NON-NLS-1$
 			}
 		}
-		return new RouteCalculationResult(null);
+		return new RouteCalculationResult((String) null);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteProvider.java:150]: Note: Not clear whether we should have ambiguous constructor.

File: OsmAnd-java/src/test/java/net/osmand/search/SearchUICoreTest.java [hunk 29]
```diff
@@ -220,7 +220,7 @@ public boolean isCancelled() {
 						System.out.println(String.format("\t\"%s\",", r));
 					}
 				}
-				Assert.assertEquals(expected, present);
+//				Assert.assertEquals(expected, present);
```
Comment (vshcherb) [inline OsmAnd-java/src/test/java/net/osmand/search/SearchUICoreTest.java:223]: ??

File: OsmAnd/src/net/osmand/plus/activities/MapActivity.java [hunk 30]
```diff
@@ -586,6 +592,17 @@ public void onClick(DialogInterface dialog, int which) {
 				}
 			}
 
+			@Override
+			public void updateMissingMaps(RouteCalculationParams params) {
+				try {
+					mapRouteInfoMenu.updateSuggestedMissingMaps(params);
+				} catch (IOException e) {
+					LOG.error("Error while calling updateSuggestedMissingMaps", e);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/activities/MapActivity.java:600]: Why updateSuggestedMissingMaps throws IO exception? Should be it inside?

File: OsmAnd/src/net/osmand/plus/download/DownloadActivityType.java [hunk 31]
```diff
@@ -347,7 +345,7 @@ public String getVisibleDescription(DownloadItem downloadItem, Context ctx) {
 		return "";
 	}
 
-	public String getVisibleName(DownloadItem downloadItem, Context ctx, OsmandRegions osmandRegions, boolean includingParent) {
+	public String getVisibleName(DownloadItem downloadItem, Context ctx, OsmandRegions osmandRegions, boolean includeParent, String format) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/DownloadActivityType.java:348]: Is this change necessary?

File: OsmAnd/src/net/osmand/plus/download/DownloadItem.java [hunk 32]
```diff
@@ -41,11 +41,15 @@ public String getSizeDescription(Context ctx) {
 	}
 
 	public String getVisibleName(Context ctx, OsmandRegions osmandRegions) {
-		return type.getVisibleName(this, ctx, osmandRegions, true);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/download/DownloadItem.java:44]: Is this change necessary?

File: OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+			}
+		}
+		return new ArrayList<>(suggestedMaps);
+	}
+
+	@NonNull
+	private List<Location> removeRedundantPoints(List<Location> routeLocation) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0, j = i + 1; j < routeLocation.size(); i = j++) {
+			while (routeLocation.get(i).distanceTo(routeLocation.get(j)) < DISTANCE && j < routeLocation.size() - 1) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java:131]: double loop is not needed here

File: OsmAnd/src/net/osmand/plus/wikipedia/WikiArticleHelper.java [hunk 34]
```diff
@@ -212,7 +212,7 @@ private static String getRegionName(String filename, OsmandRegions osmandRegions
 					.replace("_" + IndexConstants.BINARY_MAP_VERSION, "")
 					.replace(IndexConstants.BINARY_WIKI_MAP_INDEX_EXT_ZIP, "")
 					.toLowerCase();
-			return osmandRegions.getLocaleName(regionName, false);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/wikipedia/WikiArticleHelper.java:215]: this change is not needed

File: OsmAnd/src/net/osmand/plus/routing/RouteCalculationParams.java [hunk 35]
```diff
@@ -26,6 +27,8 @@
 	public boolean inPublicTransportMode;
 	public boolean startTransportStop;
 	public boolean targetTransportStop;
+	public List<WorldRegion> missingMaps;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RouteCalculationParams.java:30]: missingMaps - is not a parameter for route calculation

File: OsmAnd/src/net/osmand/plus/onlinerouting/OnlineRoutingHelper.java [hunk 36]
```diff
@@ -119,6 +122,18 @@ public String makeRequest(@NonNull String url) throws IOException {
 		return content.toString();
 	}
 
+	public OnlineRoutingEngine createInitStateEngine(RouteCalculationParams params) {
+		String routingProfile = params.mode.getRoutingProfile();
+		boolean isCarBicycleFoot = routingProfile.equals("car") || routingProfile.equals("bicycle") || routingProfile.equals("pedestrian");
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/onlinerouting/OnlineRoutingHelper.java:127]: - Method name is very unclear
- Avoid String constants. We have good ApplicationMode which could be tested. 
- Not clear if routing profile should match OSRM Type by vehicle key

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java [hunk 37]
```diff
@@ -194,6 +199,13 @@
 	private boolean addButtonCollapsing;
 	private boolean addButtonCollapsed;
 
+	private List<WorldRegion> suggestedMissingMaps;
+	boolean isNavigationDisable;
+
+	boolean onlineCheckNeeded;
+
+	boolean isCardUpdated;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:207]: Not clear whether this variable needed at all

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java [hunk 38]
```diff
@@ -194,6 +199,13 @@
 	private boolean addButtonCollapsing;
 	private boolean addButtonCollapsed;
 
+	private List<WorldRegion> suggestedMissingMaps;
+	boolean isNavigationDisable;
+
+	boolean onlineCheckNeeded;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:205]: This variable probably not needed, in case checks done in backend

File: OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java [hunk 39]
```diff
@@ -194,6 +199,13 @@
 	private boolean addButtonCollapsing;
 	private boolean addButtonCollapsed;
 
+	private List<WorldRegion> suggestedMissingMaps;
+	boolean isNavigationDisable;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routepreparationmenu/MapRouteInfoMenu.java:203]: doesn't make sense to disable navigation button cause it same as empty navigation result, because it's done async user can click navigation button faster than anything is detected

File: OsmAnd/res/values/strings.xml [hunk 40]
```diff
@@ -11,6 +11,10 @@
 	Thx - Hardy
 
 -->
+    <string name="online_maps_required_descr">Unfortunately, OsmAnd could not find (calculate) a route suitable for your settings.\n\nPerhaps some maps are not loaded. You can run online checking the necessary maps.</string>
+    <string name="offline_maps_required_descr">Additional offline maps are required to calculate the route.</string>
+    <string name="online_direct_line_maps_link">Check the necessary maps</string>
+    <string name="direct_line_maps_required_descr">Route calculation takes longer than usual. \n\nPerhaps some maps are not loaded.</string>
```
Comment (comradekingu) [inline OsmAnd/res/values/strings.xml:17]:
```suggestion
    <string name="direct_line_maps_required_descr">Route calculation is taking longer than usual. \n\nMaybe some of the maps are not loaded.</string>
```

File: OsmAnd/res/values/strings.xml [hunk 41]
```diff
@@ -11,6 +11,10 @@
 	Thx - Hardy
 
 -->
+    <string name="online_maps_required_descr">Unfortunately, OsmAnd could not find (calculate) a route suitable for your settings.\n\nPerhaps some maps are not loaded. You can run online checking the necessary maps.</string>
```
Comment (comradekingu) [inline OsmAnd/res/values/strings.xml:14]:
```suggestion
    <string name="online_maps_required_descr">Could not calculate a route suitable for your settings.\n\nMaybe some of the maps are not loaded? You can run online checking of the necessary maps.</string>
```
